core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,142 @@
+package ca.jrvs.apps.grep;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import org.apache.log4j.BasicConfigurator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JavaGrepImp implements JavaGrep {
+
+  final Logger logger = LoggerFactory.getLogger(JavaGrep.class);
+
+  private String regex;
+  private String rootPath;
+  private String outFile;
+
+  public static void main(String[] args) {
+    if (args.length != 3) {
+      throw new IllegalArgumentException("USAGE: JavaGrep regex rootPath outFile");
+    }
+
+    BasicConfigurator.configure();
+
+    JavaGrepImp javaGrep = new JavaGrepImp();
+    javaGrep.setRegex(args[0]);
+    javaGrep.setRootPath(args[1]);
+    javaGrep.setOutFile(args[2]);
+
+    try {
+      javaGrep.process();
+    } catch (Exception ex) {
+      javaGrep.logger.error(ex.getMessage(), ex);
```

Comment:
javaGrep.logger.error("meaninful message", ex);
      
      you want to avoid double logging the same message